src/aspire/orientation/commonline_sync.py

```diff
@@ -1,31 +1,411 @@
 import logging
+import numpy as np
 
-from aspire.orientation import Orient3D
+from scipy.spatial.transform import Rotation
+import scipy.sparse as sparse
+
+from aspire.orientation import CLOrient3D
+from aspire.utils import ensure
 
 logger = logging.getLogger(__name__)
 
 
-class CommLineSync(Orient3D):
+class CommLineSync(CLOrient3D):
     """
-    Define a class to estimate 3D orientations using Synchronization described as below:
-    Y. Shkolnisky, and A. Singer, Viewing Direction Estimation in Cryo-EM Using Synchronization,
+    Define a class to estimate 3D orientations using synchronization matrix and voting method.
+
+    The related publications are listed as below:
+    Y. Shkolnisky, and A. Singer,
+    Viewing Direction Estimation in Cryo-EM Using Synchronization,
     SIAM J. Imaging Sciences, 5, 1088-1110 (2012).
 
+    A. Singer, R. R. Coifman, F. J. Sigworth, D. W. Chester, Y. Shkolnisky,
+    Detecting Consistent Common Lines in Cryo-EM by Voting,
+    Journal of Structural Biology, 169, 312-322 (2010).
     """
-    def __init__(self, src):
+    def __init__(self, src, n_rad=None, n_theta=None):
         """
-        constructor of an object for estimating 3D orientations
+        Initialize an object for estimating 3D orientations using synchronization matrix
+
+        :param src: The source object of 2D denoised or class-averaged images with metadata
+        :param n_rad: The number of points in the radial direction
+        :param n_theta: The number of points in the theta direction
         """
-        pass
+        super().__init__(src, n_rad=n_rad, n_theta=n_theta)
+        self.syncmatrix = None
 
-    def estimate(self):
+    def estimate_rotations(self):
         """
-        perform estimation of orientations
+        Estimate orientation matrices for all 2D images using synchronization matrix
         """
-        pass
+        if self.syncmatrix is None:
+            self.syncmatrix_vote()
+
+        S = self.syncmatrix
+        sz = S.shape
+        ensure(sz[0] == sz[1], 'syncmatrix must be a square matrix.')
+        ensure(sz[0] % 2 == 0, 'syncmatrix must be a square matrix of size 2Kx2K.')
+
+        n_img = sz[0] // 2
+
+        # S is a 2Kx2K matrix (K=n_check), containing KxK blocks of size 2x2.
+        # The [i,j] block is given by [r11 r12; r12 r22], where
+        # r_{kl}=<R_{i}^{k},R_{j}^{l}>, k,l=1,2, namely, the dot product of
+        # column k of R_{i} and columns l of R_{j}. Thus, given the true
+        # rotations R_{1},...,R_{K}, S is decomposed as S=W^{T}W where
+        # W=(R_{1}^{1},R_{1}^{2},...,R_{K}^{1},R_{K}^{2}), where R_{j}^{k}
+        # the k column of R_{j}. Therefore, S is a rank-3 matrix, and thus, it
+        # three eigenvectors that correspond to non-zero eigenvealues, are linear
+        # combinations of the column space of S, namely, W^{T}.
+
+        # Extract three eigenvectors corresponding to non-zero eigenvalues.
+        d, v = sparse.linalg.eigs(S, 10)
```

Comment:
It looks that this part tries to repeat literally from Matlab code. It finds the top 10 eigenvectors based on the largest magnitude and then find the largest 3 based on the real values of eigenvalues.  In some cases it might be different to use `d, v = sparse.linalg.eigs(S, 3, which='LR')`. Since the result from simulation example is reasonable, I will keep as it is.

Reply:
Okay.  Probably we should 1) write that down or link to a doc for why we do this dance 2) ask the author of that method what the motivation was here and if there might be something more effective.

For example, the largest three values are usually the most accurate with accuracy decreasing rapidly...  Having done a lot of eigen-stuff in the past, I have a lot of reservations about lower order eigenvalues/vectors in systems with a lot of noise.

If we just need three, you might get the first 6 (instead of 10) since they should be coming in pairs if they are imaginary. Actually, hrmmm, if they are imaginary, this sorting scheme would maybe be degenerate, your top two magnitude would be a conjugate pair...  Reading the comment, it is saying S = W.T * W, so I take it S is symmetric then? If real symmetric then eigenvalues should be real. If that is true we almost certainly want to use symmetric solver, maybe this one, and no reason to deal with imaginary parts (they'd be numerical noise then).

https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.eigsh.html#scipy.sparse.linalg.eigsh

@janden , what do you know of this? Does this method originate with PR or Tel Aviv, I think this bit is worth getting right. I actually thought about math today, how about that :D.

Reply:
I skimmed the section about this matrix in the associated paper. Indeed, S is symmetric by construction, clearly stated.  That potentially has a lot of implications.

Yoel was the first author, so I dropped him a message. It isn't common to have such close relationships with authors IME, so we should take advantage of that... (this is perhaps one benefit of academic software development, for all it's faults).

Reply:
Talked with Yoel. We can improve this section. I think the original matlab code was doing something slightly smarter and it was lost in translation.  

From my chat:
1) We can and probably should use a symmetric solver. Follows eigenvalues should be real.
2) The original code asked for 10 eigenvalues so they could print/review the gaps, not because of the sorting etc.
3) We should expect first three values to be close, gap between third and fourth.
4) We should be logging the eigenvalues returned for inspection.  We can probably just get away with first 4 assuming sorted.
5) There may not be a good heuristic for automatically testing this gap at this time for real data sets. For clean/contrived data sets we can.

Lets discuss this today. Hope that helps.

Reply:
For this PR we discussed:
 change to `eigsh` 
 removing the `np.real` calls (they are not necessary)
 logging the eigenvalues.

 I will create an issue for the other topic (how many values, writing out diagnostics etc).

Reply:
Done in commit 7d3af4d.

Reply:
#211 and #212 cover the other points.  Thanks.